pkg/tecdsa/dkg/states.go

```diff
@@ -163,29 +166,26 @@ func (tros *tssRoundOneState) ActiveBlocks() uint64 {
 }
 
 func (tros *tssRoundOneState) Initiate(ctx context.Context) error {
-	// The ctx instance passed as Initiate argument is scoped to the lifetime
-	// of the current state. However, the Initiate method is blocking and the
-	// ctx instance is cancelled properly only after Initiate returns. Because
-	// of that, we cannot use ctx as round timeout signal as Initiate would
-	// hang forever if something goes wrong. To avoid such a resource leak,
-	// we set a round timeout based on state block duration and an average
-	// block time. The exact duration doesn't need to be super-accurate because
-	// if the timeout is hit, the execution will fail anyway. We just want
-	// to give enough time for round computation and make sure the round
-	// terminates regardless of the result.
-	stateBlocks := tros.DelayBlocks() + tros.ActiveBlocks()
-	stateDuration := 15 * time.Second * time.Duration(stateBlocks)
-	roundCtx, roundCtxCancel := context.WithTimeout(ctx, stateDuration)
-	defer roundCtxCancel()
-
-	message, err := tros.member.tssRoundOne(roundCtx)
-	if err != nil {
-		return err
-	}
+	// TSS computations can be time-consuming and can exceed the current
+	// state's time window. The ctx parameter is scoped to the lifetime of
+	// the current state so, it can be used as a timeout signal. However,
+	// that ctx is cancelled upon state's end only after Initiate returns.
+	// In order to make that working, Initiate must trigger the computations
+	// in a separate goroutine and return before the end of the state.
+	go func() {
```

Comment:
Not in this way but yeah, we can use a single channel for the outcome. See 5143c7d5

Reply:
The code snippet was a copy-paste from the previous version, I did not suggest implementing it this way, just wanted to explain where is the mentioned _We later check if the value read from `tros.errChan` is not-nil_ :)

I think the new version still leaks. If there is an error in `tros.member.tssRoundOne(ctx)` or in `tros.channel.Send(ctx, message);` the goroutine returns without closing `tros.outcomeChan`. It is closed only on success. Instead of closing it manually, shouldn't we `defer` in the first line of the goroutine code?

Reply:
It is ok to leave channels open. There are no leaks in this case as those channels are garbage collected once they become unreachable ([source](https://tip.golang.org/doc/gc-guide#Tracing_Garbage_Collection)) It's only necessary to close a channel when it is important to tell the receiving goroutines that all data have been sent. In this case, closed channels mean success so we shouldn't close them in all cases.

Reply:
I keep forgetting about it...